### ISSUE TITLE
Reference manual does not build (raises fatal warnings) with Sphinx 3.

### ISSUE THREAD

**Issue by Janno** [coq/coq#12332]
#### Description of the problem

When building refman-html (via `dune` or `make`) the build is stopped by the following fatal warning:
```
Warning, treated as error:
/home/janno/src/coq/_build/default/doc/sphinx/proof-engine/ltac2.rst:665:duplicate token description of coq:ltac2_term       , other instance in proof-engine/ltac2
```

#### Coq Version
master+changes from #12329.

### Python package versions:

```
antlr4-python3-runtime==4.8
beautifulsoup4==4.9.0
pexpect==4.8.0
sphinx-rtd-theme==0.4.3
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-bibtex==1.0.0
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==1.0.3
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.4
```

**Comment by Janno** [comment]
Small addendum: looking at /tmp/warn.log after running `SPHINXWARNOPT="-w/tmp/warn.log" dune build @refman-html` gave me (amongst many other lines) these two here that might be helpful:
```
/home/janno/src/coq/_build/default/doc/sphinx/proof-engine/ltac2.rst:665: WARNING: duplicate token description of coq:ltac2_term       , other instance in proof-engine/ltac2
/home/janno/src/coq/_build/default/doc/sphinx/proof-engine/ltac2.rst:916: WARNING: duplicate token description of coq:ltac2_term , other instance in proof-engine/ltac2
```

Note the whitespace! Could that be relevant?

**Comment by Zimmi48** [comment]
@Janno Could you please also report your version of Sphinx? (Run `sphinx-build --version`)

**Comment by Janno** [comment]
`sphinx-build 3.0.3`

**Comment by Zimmi48** [comment]
OK, so this seems to be a newly introduced bug in Sphinx (or a Coq bug that only appears on recent Sphinx versions). Could you check if downgrading to 2.4.4 makes the issue disappear? I'll try to investigate whether I can get a minimal reproducible example for this issue, but we may have to document that Coq requires Sphinx < 3.0.

**Comment by Zimmi48** [comment]
I retract what I said of this being a new bug. It's just that the Ltac2 chapter abuses `productionlist` by redefining over and over the `ltac2_term` token. I think that the following patch should be enough as a temporary workaround until @jfehrle rewrites the grammar in this chapter to use `prodn`:

```diff
diff --git a/doc/sphinx/proof-engine/ltac2.rst b/doc/sphinx/proof-engine/ltac2.rst
index 1e35160205..2e5a77d4d4 100644
--- a/doc/sphinx/proof-engine/ltac2.rst
+++ b/doc/sphinx/proof-engine/ltac2.rst
@@ -662,7 +662,7 @@ Match over terms
 Ltac2 features a construction similar to Ltac1 :n:`match` over terms, although
 in a less hard-wired way.
 
-.. productionlist:: coq
+.. productionlist:: coq_2
    ltac2_term       : match! `ltac2_term` with `constrmatching` .. `constrmatching` end
                     : lazy_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
                     : multi_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
@@ -692,7 +692,7 @@ Match over goals
 Similarly, there is a way to match over goals in an elegant way, which is
 just a notation desugared at parsing time.
 
-.. productionlist:: coq
+.. productionlist:: coq_3
    ltac2_term             : match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
                     : lazy_match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
                     : multi_match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
@@ -907,13 +907,13 @@ a Ltac1 expression, and semantics of this quotation is the evaluation of the
 corresponding code for its side effects. In particular, it cannot return values,
 and the quotation has type :n:`unit`.
 
-.. productionlist:: coq
+.. productionlist:: coq_4
   ltac2_term : ltac1 : ( `ltac_expr` )
 
 Ltac1 **cannot** implicitly access variables from the Ltac2 scope, but this can
 be done with an explicit annotation on the :n:`ltac1` quotation.
 
-.. productionlist:: coq
+.. productionlist:: coq_5
   ltac2_term : ltac1 : ( `ident` ... `ident` |- `ltac_expr` )
 ```

**Comment by jfehrle** [comment]
@Janno Add `SPHINXWARNERROR=0` on your make command line to let the build complete on your system.

**Comment by Zimmi48** [comment]
@jfehrle `SPHINXWARNOPT="-w/tmp/warn.log" dune build @refman-html` does the same thing for Dune.

**Comment by Zimmi48** [comment]
After investigating, it seems that there are two sources of fatal warnings when building with Sphinx 3:
- tokens that are redefined in multiple `productionlist` in the same production group (only happens in the Ltac2 chapter);
- tokens that are defined in `productionlist` do not seem to be referable from a `tacn` or `cmd`.

The second problem might be considered a bug of coqrst (cc @cpitclaudel). That being said, since we have made good progress toward removing most `productionlist` from the reference manual, a way to provide Sphinx 3 compatibility would be to finish the conversion (Ltac and Ltac2 are the large remaining pieces of work, but Ltac could be achieved for 8.12+beta1 and Ltac2 for 8.12.0, while the rest of the `productionlist` could be converted quickly without necessarily fixing the syntax at the same time).

**Comment by Zimmi48** [comment]
My proposal for "resolving" this bug for 8.12+beta1 is to document in `doc/README.md` and the changelog that the manual is known not to be compatible with Sphinx 3. For 8.12.0, we might be able to just get rid of the remaining `productionlist`.

@cpitclaudel: WDYT?

**Comment by Zimmi48** [comment]
I'm going to implement the strategy I proposed.

**Comment by Zimmi48** [comment]
#12459 documents the incompatibility. Keeping this issue open (until compatibility is fixed) but now targeting 8.12.0.
